airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py

```diff
@@ -0,0 +1,257 @@
+        headers: Optional[Mapping[str, str]] = None,
+        params: Optional[Mapping[str, str]] = None,
+        json: Optional[Mapping[str, Any]] = None,
+        data: Optional[Union[str, Mapping[str, Any]]] = None,
+    ):
+        if dedupe_query_params:
+            query_params = self._dedupe_query_params(url, params)
+        else:
+            query_params = params or {}
+        args = {"method": http_method, "url": url, "headers": headers, "params": query_params}
+        if http_method.upper() in BODY_REQUEST_METHODS:
+            if json and data:
+                raise RequestBodyException(
+                    "At the same time only one of the 'request_body_data' and 'request_body_json' functions can return data"
+                )
+            elif json:
+                args["json"] = json
+            elif data:
+                args["data"] = data
+        prepared_request: requests.PreparedRequest = self._session.prepare_request(requests.Request(**args))
+
+        return prepared_request
+
+    def _send_with_retry(self, request: requests.PreparedRequest, request_kwargs: Optional[Mapping[str, Any]] = None) -> requests.Response:
+        """
+        Backoff package has max_tries parameter that means total number of
+        tries before giving up, so if this number is 0 no calls expected to be done.
+        But for this class we call it max_REtries assuming there would be at
+        least one attempt and some retry attempts, to comply this logic we add
+        1 to expected retries attempts.
+        """
+        if self._backoff_strategy.max_retries is not None:
+            max_tries = max(0, self._backoff_strategy.max_retries) + 1
+
+        user_backoff_handler = user_defined_backoff_handler(max_tries=max_tries, max_time=self._backoff_strategy.max_time)(self._send)
+        backoff_handler = default_backoff_handler(
+            max_tries=max_tries, max_time=self._backoff_strategy.max_time, factor=self._backoff_strategy.retry_factor
+        )
+        # backoff handlers wrap _send, so it will always return a response
+        response = backoff_handler(user_backoff_handler)(request, request_kwargs)
+
+        return response
+
+    def _send(self, request: requests.PreparedRequest, request_kwargs: Optional[Mapping[str, Any]] = None) -> requests.Response:
+
+        self._logger.debug(
+            "Making outbound API request", extra={"headers": request.headers, "url": request.url, "request_body": request.body}
+        )
+
+        if request_kwargs is None:
+            request_kwargs = {}
+
+        try:
+            response: requests.Response = self._session.send(request, **request_kwargs)
+            self._response_decoder.validate_response(response)
+        except requests.RequestException as exc:
+            response_action = ResponseAction.RETRY
+            failure_type = FailureType.transient_error
+            response_error_message = f"Request to {request.url} failed with exception: {exc}"
```

Comment:
Can we add the method here? Sometimes, we get the same URL with different methods hence this would be useful to distinguish which one is which